### ISSUE TITLE
Provide documentation for Glycan Hightlight using SVG

### ISSUE THREAD

**Issue by ReneRanzinger** [glygener/glygen-issues#496]
Provide information and documentation about the usage of SVG for glycan feature highlighting. We will use this to evaluate what kind of highlighting we want to support in GlyGen.

**Blocker for**:
* #497

**Comment by edwardsnj** [comment]
Here is the two demo URLs I showed: 
https://glymage.glyomics.org/demo/demo1.html
https://glymage.glyomics.org/demo/demo2.html

**Comment by edwardsnj** [comment]
Example JSON annotation files: [Example JSON Annotation Files.zip](https://github.com/glygener/glygen-issues/files/12599359/Example.JSON.Annotation.Files.zip)

**Comment by edwardsnj** [comment]
@ReneRanzinger  The SVG highlighting and annotation infrastructure has been pushed into the docker image and publicly released. URLs above have been updated to point to the publicly deployed location - and everything seems to be working. Enzyme annotations now include the link to the monosaccharide. Users can provide their own javascript functions for determining the SVG and JSON urls (demonstrated in demo2.html). I will push GlyGen subset PNG, SVG, and JSON files for snfg + extended to the GlycanData export master branch shortly.

**Comment by ReneRanzinger** [comment]
It would be nice to support the "Sandbox style" highlighting:
* set CSS `{opacity: 0.3;}` for all elements (including linkages and linkage labels)
* set CSS `{opacity: 1;}` for the highlighted elements

Nice to have feature.

**Comment by edwardsnj** [comment]
See: https://edwardslab.bmcb.georgetown.edu/glymagedev/demo/demo3.html

This is about as good I can figure at at this time. The GalNAc with opacity 0.3 allows the link line below to be visible. This is due to the way the SVG file is structured and how opacity works for grouped (and not grouped) objects. It is not inconceivable that there is a resolution to this, and potentially I may be able to manipulate the SVG file structure a bit if there is a good way to do that, but...

This is not for the remote clickable case, but I will add the image class and link info class logic there too. 

Also needed to imitate the sandbox highlighting style, is the ability to turn the background grey? and have a click on the background clear everything. Work still needed...

I will use demo3.html as the place to experiment and demonstrate this stuff (not up permanently, but so you can see it and I can experiment with it).

**Comment by ReneRanzinger** [comment]
The grey background and the clear click are not high priority. We can also have a "Clear" button outside the SVG and that works well. Now that I see this I remember the line (linkage) behind the shape (residue) issue. I think Wills solution was to have a white copy of the shape between the "color shape" and the linkage line. The white copy does not get the opacity set. The "color shape" has the white copy shining through but the non-transparent white shape hides the ends of the linkage lines that are below the shapes. Hope that makes sense.

**Comment by edwardsnj** [comment]
It does make sense. I already experimented along those lines to try achieve something similar and couldn't figure it out (given the SVG file as it currently exists). The current way that the lines and shapes are structured seems (at least to my understanding) to preclude this approach. I am no SVG/CSS guru. Please feel free to tinker to solve the problem. 

here is a helpful stackoverflow post explaining some of the issues:

https://stackoverflow.com/questions/15825758/can-the-opacity-translucence-of-an-svg-group-be-adjusted-together

**Comment by ReneRanzinger** [comment]
That works nicely to have the lines not shining through but you can not get one of the inner elements to not be transparent. Its like laying a semi-transparent square on top of everything. We tested that, too. If you want to use that you would have to dynamically change the structure of the SVG and move the elements (that are supposed to be not transparent) outside of the `<g>` that you assigned the lower opacity. 

For Wills solution: whenever you create a shape in the SVG:
```
<g id="r-1:2" data.residueanomericstate="b" data.residuechirality="D" data.residueindex="1" data.residueisreducingend="true" data.residuename="GlcNAc" data.residueringsize="p" data.type="Monosaccharide" style="fill:white; text-rendering:optimizeSpeed; stroke:white;" class="glymagesvg_outline_red">
      <rect height="22" style="clip-path:url(#clipPath1:171c3393e62159e41a243923d4ae213); stroke:none;" width="22" x="134" y="30"></rect>
      <rect id="rect1" height="22" style="clip-path:url(#clipPath1:171c3393e62159e41a243923d4ae213); fill:rgb(0,114,188); stroke:none;" width="22" x="134" y="30"></rect>
      <rect height="22" style="clip-path:url(#clipPath1:171c3393e62159e41a243923d4ae213); fill:none; stroke:black;" width="22" x="134" y="30"></rect>
      <rect height="22" style="fill:none; stroke:black;" width="22" x="134" y="30"></rect>
    <title>GlcNAc</title></g>
```

Right before that you would create another shape in its own `<g>` with the same size, coordinates, no outline and white. In terms of order it will come before the actual shape but after the linkage lines. When you assign opacity, you assign it to all shapes and lines individually except for the extra white shapes. But it will require to change the SVG generation.

**Comment by edwardsnj** [comment]
OK, I looked at Will's and your example above, and have been able to make this work. I dynamically clone nodes in the SVG after I pull it from the server. *eye roll*. Ultimately I may change the SVG generating code I have implemented in GlycanBuilder2 to avoid doing this on the fly. I have demonstrated this in the demo3.html link above. I will push this functionality into the public release next week.

**Comment by ReneRanzinger** [comment]
I like this version better than the outline but that is personal preference. I noticed an issues. On the hover example the white shape seems to be missing (the line shines through the GalNAc).

**Comment by edwardsnj** [comment]
I have completed the work to support highlighting as requested in the same manner as for the sandbox. The CSS to achieve this is now part of the glymage docker container and the demo3 example is also now packaged in the public glymage server. Additional parameters to support sizing and various "click" modes also implemented. Reset/clear by clicking on the background of the image. The master branch of the glycan data repository has updated GlyGen PNG, SVG, and JSON files for Robel to serve.

**Comment by sujeetvkulkarni** [comment]
@edwardsnj Can you please point me to the repository from where I can get current JSON, SVG and JavaScript files for Glycan Highlight widget. Are the JSON files attached above "Example JSON annotation files" current?

**Comment by edwardsnj** [comment]
@sujeetvkulkarni The images*.tbz.* files from [Glycan Data Export](https://github.com/glygen-glycan-data/PyGly/tree/master/smw/glycandata/export) contain the current SVG, JSON (and PNG) files for GlyGen. 

The Javascript and CSS files are in the repository:

  https://github.com/glygen-glycan-data/APIFramework

Specifically:

  https://github.com/glygen-glycan-data/APIFramework/blob/main/src/Application/Glymage/js/glymagesvg.js

and 

  https://github.com/glygen-glycan-data/APIFramework/blob/main/src/Application/Glymage/css/glymagesvg.css

Hope this helps...

**Comment by sujeetvkulkarni** [comment]
@edwardsnj Thanks a lot, this certainly helps.

**Comment by edwardsnj** [comment]
Glymage SVG highlighting released with version 2.3 of the portal, all current issues resolved.
